[improve][pulsar-io] Added support for generic record and raw JSON string schemas to CassandraSink

david-streamlio opened the pull request:
### Motivation

The current implementation of the Cassandra Sink connector only supported a single schema type (key, string). This is not useful for production. So I modified the code to be able to support any schema type in Cassandra. 

### Modifications

Added classes that interrogate the database to determine the schema type at runtime. I also added a framework that will extract the values from the supported incoming schema types (GenericRecord, and String) using the table metadata.

### Verifying this change

- [ x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change added tests and can be verified as follows:

   *Added integration tests for testing against a Cassandra database*
 
### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API: (no)
  - The schema: (yes)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: ( no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ x] `doc-required` 
(Your PR needs to update docs and you will update later)

david-streamlio committed:
Added GenericRecord support

david-streamlio committed:
Added support for raw JSON String

david-streamlio committed:
Added RecordWrapper interface

david-streamlio commented:
@tspannhw  can you review and upvote?

david-streamlio commented:
/pulsarbot run-failure-checks

david-streamlio commented:
/pulsarbot ready-to-test

eolivelli commented on pulsar-io/cassandra/pom.xml:
```diff
@@ -51,6 +51,39 @@
+    <dependency>
+      <groupId>org.apache.pulsar</groupId>
+      <artifactId>pulsar-functions-local-runner-original</artifactId>
+      <version>2.11.0-SNAPSHOT</version>
```
please use project.version

eolivelli commented on pulsar-io/cassandra/src/main/java/org/apache/pulsar/io/cassandra/util/CassandraConnector.java:
```diff
@@ -0,0 +1,135 @@
+public class CassandraConnector {
+
+    private Cluster cluster;
+    private Session session;
```
we have to handle concurrent access properly to these fields

eolivelli commented on pulsar-io/cassandra/src/main/java/org/apache/pulsar/io/cassandra/util/StringRecordWrapper.java:
```diff
@@ -0,0 +1,47 @@
+
+    public StringRecordWrapper(String jsonString) {
+        super(jsonString);
+        valuesMap = new Gson().fromJson(jsonString,
```
I think that we can cache this "new Gson()" instance in a static field

eolivelli commented on pulsar-io/cassandra/src/main/java/org/apache/pulsar/io/cassandra/util/TableMetadataProvider.java:
```diff
@@ -0,0 +1,102 @@
+
+package org.apache.pulsar.io.cassandra.util;
+
+import com.datastax.driver.core.*;
```
please do not use star imports

eolivelli commented on pulsar-io/cassandra/src/test/java/org/apache/pulsar/io/cassandra/CassandraSinkExec.java:
```diff
@@ -0,0 +1,98 @@
+        File file = new File(classLoader.getResource(CONFIG_FILE).getFile());
+
+        Yaml yaml = new Yaml();
+        configs = yaml.load(new FileInputStream(file));
```
please close the stream

eolivelli commented on pulsar-io/cassandra/src/test/java/org/apache/pulsar/io/cassandra/CassandraSinkExec.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private static final void sendData() throws InterruptedException {
+        TimeUnit.SECONDS.sleep(10);
```
do we need to sleep a fixed amount of time ?

david-streamlio replied:
The only reason I sleep 10 seconds, is so I can easily visualize the data as it is published to the system. The delay allows me time to execute a query against Cassandra in order to detect the new records.

eolivelli commented on pulsar-io/cassandra/src/test/java/org/apache/pulsar/io/cassandra/producers/InputTopicProducerThread.java:
```diff
@@ -0,0 +1,83 @@
+            try {
+                getProducer().newMessage().key(getKey()).value(getValue()).send();
+            } catch (PulsarClientException e) {
+                e.printStackTrace();
```
please use logger

eolivelli reviewed (changes requested):
Great work, thanks

I left some comments, please take a look

david-streamlio committed:
fixed issue from code review

david-streamlio commented:
@eolivelli I have made the requested changes, can you PTAL when you get a chance?  Thank!

david-streamlio commented:
@eolivelli , Can you please take a look at this when you get the chance? Thanks again!

david-streamlio commented:
@eolivelli Can I please get some feedback on these changes I made in response to your initial feedback? Thanks again for the review, I really appreciated it.

tisonkun committed:
Merge branch 'master' into cassandra-sink-fix

tisonkun committed:
license header

Signed-off-by: tison <wander4096@gmail.com>

tisonkun reviewed (dismissed):
@david-streamlio Please fix the compilation failure. It seems that you use JUnit while Pulsar use TestNG as the test platform.

I've pushed a merge commit and fixing for license header to your remote branch so be aware to `git pull` before working on it.

BTW, please try to expand all star import.

david-streamlio committed:
Switched to TestNG framework

david-streamlio commented:
/pulsarbot run-failure-checks

david-streamlio commented:
@tisonkun @eolivelli I would appreciate another review when you have the time.

eolivelli reviewed (approved):
+1

lhotari committed:
Merge remote-tracking branch 'origin/master' into cassandra-sink-fix

lhotari committed:
Revert unnecessary changes to pom.xml files that ignore cassandra.version property

lhotari committed:
Fix checkstyle

lhotari committed:
Replace Gson usage with Jackson

lhotari committed:
Remove version for beanutils

lhotari committed:
Remove unnecessary dependency

lhotari committed:
Add Testcontainers dependency

lhotari committed:
Use real Cassandra with Testcontainers

lhotari reviewed (approved):
LGTM. There doesn't seem to be integration tests, but those could be added later.

lhotari committed:
Revert renaming of CassandraAbstractSink

lhotari reviewed (changes requested):
The behavior of the existing CassandraStringSink shouldn't be modified since it breaks backwards compatibility. Breaking existing behavior will be surprising for existing users. That's why it's better to add the new behavior only when specific configuration is used.

david-streamlio commented:
Closing this in favour of two PRs against `apache/pulsar-connectors`, where the connector now lives:

- **apache/pulsar-connectors#136** — authenticated clusters (`userName` / `password`)
- **apache/pulsar-connectors#137** — the `cassandra-generic-record` and `cassandra-json` sinks

**Why it moved rather than being updated here.** PIP-465 removed the IO connectors from this
repository (#25408, merged 2026-03-27), a day after `apache/pulsar-connectors` was initialised with
the Cassandra connector as it then stood. The connector migrated cleanly; this PR, still open on top
of it, did not. Every file here targets `pulsar-io/cassandra/**`, which no longer exists on `master`,
plus a `pom.xml`, and none of those survive the Gradle migration either. Rebasing produces
delete/modify conflicts on all 23 files with nowhere to resolve them to, so the change had to be
ported rather than rebased. Apologies for leaving it sitting this long in the first place.

**Why two PRs.** Re-reading this one, it was doing two independent things under a title that mentions
only one. The Cassandra sink cannot connect to a cluster requiring authentication at all — there are
no credential fields and `CassandraAbstractSink` builds its `Cluster` with contact points only — and
that fix has nothing to do with schema mapping. It is four files and reviewable on its own, so it
should not wait behind a design discussion about field-to-column mapping. That is #136. The
schema-mapping work is #137, stacked on it.

**@lhotari — your change request is addressed, and I took it further than asked.** You wrote:

> The behavior of the existing CassandraStringSink shouldn't be modified since it breaks backwards
> compatibility. Breaking existing behavior will be surprising for existing users. That's why it's
> better to add the new behavior only when specific configuration is used.

You were right, and re-reading the diff the damage was wider than the line you flagged. This PR also
dropped the `@Connector` annotation that registers the `cassandra` sink, changed the input from
`Record<byte[]>` to `Record<String>`, and removed the key-falls-back-to-value behaviour — on top of
reinterpreting the message as JSON. Four breaking changes to a shipped connector, for a feature that
did not need to touch it.

Rather than gate the new behaviour behind a configuration flag, #137 makes it a separate connector.
`CassandraStringSink`, `CassandraAbstractSink` and the `cassandra` entry in `pulsar-io.yaml` are
unchanged, and the new behaviour arrives as `cassandra-generic-record` (schema-carrying records) and
`cassandra-json` (raw JSON strings). An operator selects one by name or does not get it, so there is
no configuration path by which the existing sink can start behaving differently. Its diff is
additive — 934 insertions, zero deletions — and the pre-existing `CassandraStringSinkTest` passes
unmodified, which is the check that backs the compatibility claim.

One thing in #136 does touch the existing path and is worth your eye: config loading moves from
`CassandraSinkConfig.load` to `IOConfigUtils.loadWithSecrets`, so a `sensitive` password can come
from a Pulsar secret instead of plaintext config. It is the helper kinesis, redis, canal and
azure-data-explorer already use, and `CassandraStringSinkTest` passing unchanged is the evidence that
it is a drop-in for configs with no secrets — but it is a real change to that path and I would rather
point at it than have it found.

Your other note — that there were no integration tests and they could come later — is partly
answered. Between the two PRs there are Testcontainers tests for authentication, the connection, the
generated `INSERT` against real table metadata, and table-definition resolution. There is still no
end-to-end test driving either new sink through a running sink instance; that is called out in #137
as worth doing rather than left implied.

Thanks for the review. It was the right call and the design is better for it.
